### ISSUE TITLE
Remove requirement for context.Context to have a deadline

### ISSUE THREAD

**Issue by Merovius** [nats-io/nats.go#855]
## Feature Request

Currently, most functions/methods taking a `context.Context` require that context to have a deadline set, returning an error otherwise. I presume that also means they will ignore context cancellation, if not due to a timeout. That's unfortunate, as it means code might block for longer than it should - possibly *very* long, as it makes sense to have very long timeouts for pubsub-operations.

#### Use Case:

There are many use-cases which require cancellation without deadlines. For example, you might want to use [errgroup](https://golang.org/x/sync/errgroup) to do multiple concurrent things and cancel them all once the first error occurs. 

#### Proposed Change:

It is possible to asynchronously cancel network read/write operations (which I assume is the only aspect of the NATS client where cancellation would apply) by [calling `SetDeadline` with a time in the past](https://pkg.go.dev/net#Conn.SetDeadline):

> A deadline is an absolute time after which I/O operations
> fail instead of blocking. The deadline applies to all future
> **and pending** I/O, not just the immediately following call to
> Read or Write. After a deadline has been exceeded, the
> connection can be refreshed by setting a deadline in the future.
> 
> If the deadline is exceeded a call to Read or Write or to other
> I/O methods will return an error that wraps os.ErrDeadlineExceeded.
> This can be tested using errors.Is(err, os.ErrDeadlineExceeded).

So, a way to get a network connection receptive to cancellation could be something akin to

```go
// withCancel calls f, making sure that any read/write operations on
// c are cancelled when ctx is.
func withCancel(ctx context.Context, c net.Conn, f func(context.Context, net.Conn) error) error {
    // wait for the spawned goroutine to stop before returning, to prevent
    // concurrent setting of the deadline, when re-using the connection.
    wg := new(sync.WaitGroup)
    defer wg.Wait()

    // add a layer of cancellation, to always stop the spawned goroutine
    // before returning
    ctx, cancel := context.WithCancel(ctx)
    defer cancel()

    if dl, ok := ctx.Deadline(); ok {
        c.SetDeadline(dl)
    }

    wg.Add(1)
    go func() {
        defer wg.Done()
        <-ctx.Done()
        c.SetDeadline(time.Now())
    }()
    err := f(ctx, c)
    if e := ctx.Done() && e != nil && errors.Is(err, os.ErrDeadlineExceeded) {
        err = e
    }
    return err
}
```

Of course the details of how best to do this will depend on the actual client code and how it plumbs connections, but this shows all the relevant parts.

#### Who Benefits From The Change(s)?

Anyone who might use a `context.Context` that might get cancelled not based on deadlines.

#### Alternative Approaches

I'm not sure.

**Comment by wallyqs** [comment]
Hi @Merovius, could you share an example of using `WithCancel` that fails? I think the only APIs that do not support it are `FlushWithContext ` and we can fix that to automatically create a child context so that it defaults to the same timeout as `nc.Flush()`. The only thing that is not supported is passing `context.Background()` since that will throw an error:

https://github.com/nats-io/nats.go/blob/2554593026cc24f0c860c6112c5c21520b932e42/js.go#L2350-L2354

**Comment by Merovius** [comment]
@wallyqs Ah that's fair. I observed the error in a test, where I was actually using `context.Background` and extrapolated from the error message and the docs of `FlushWithContext`.

So, just to confirm, cancellation without a deadline *does* work? I grepped for `Set(Read|Write)?Deadline` and it didn't turn up any meaningful results. AFAIK the method of setting a past deadline is the only way to abort an in-progress read/write of a `net.Conn`.

**Comment by wallyqs** [comment]
Yes it would work with the blocking APIs like Request or NextMsg for example, and most of the JS context APIs too.

**Comment by Merovius** [comment]
Okay, great, it seems that it's only `FlushWithContext` then. And maybe the error message could be a bit more descriptive, if the issue is usage of `context.Background()`, not the lack of a deadline :)

**Comment by wallyqs** [comment]
Sounds good to me, ok we can fix that

**Comment by stefanvanburen** [comment]
Ran into a similar issue while testing NATS and realized that passing `context.Background()` wasn't allowed, but I'm curious why `context.Background()` is handled differently than say, `context.TODO()` or just a random `context.WithValue()` without a timeout or cancel attached? Shouldn't `context.Background()` just fall through to the lines below?:

https://github.com/nats-io/nats.go/blob/907b219666457d069b43740b9d9cf9ae234075a8/js.go#L2649-L2652
